# Data Cleaning & Preprocessing

**Project:** Medical Insurance Cost Prediction  
**Phase:** Day 17 - Data Cleaning & Preprocessing  
**Dataset:** `../data/insurance.csv`  
**Target Variable:** `charges`

This notebook inspects the raw dataset for data-quality issues and prepares it for machine learning.
No model training, train/test splitting, or feature scaling is performed here.

## 1. Load Dataset

In [1]:
import pandas as pd

df = pd.read_csv('../data/insurance.csv')
print('Dataset loaded successfully.')
print('Initial shape:', df.shape)

Dataset loaded successfully.
Initial shape: (2772, 7)


## 2. Missing-Value Analysis

In [2]:
print('Missing values per column:')
print(df.isnull().sum())
print()
print('Total missing values:', df.isnull().sum().sum())

Missing values per column:
age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

Total missing values: 0


> **Result:** No missing values found in any column.  
> No missing-value treatment was required.

## 3. Duplicate Analysis

In [3]:
print('Duplicate rows before removal:', df.duplicated().sum())
print('Row count before:', df.shape[0])

Duplicate rows before removal: 1435
Row count before: 2772


In [4]:
df = df.drop_duplicates().reset_index(drop=True)
print('Duplicate rows after removal:', df.duplicated().sum())
print('Row count after:', df.shape[0])

Duplicate rows after removal: 0
Row count after: 1337


> **Result:** 1,435 duplicate rows were found and removed.  
> Dataset reduced from **2,772 rows** to **1,337 rows**.

## 4. Data-Type Checks

In [5]:
df.dtypes

age           int64
sex          object
bmi         float64
children      int64
smoker       object
region       object
charges     float64
dtype: object

> **Result:** All numeric columns (`age`, `bmi`, `children`, `charges`) have correct numeric dtypes.  
> Categorical columns (`sex`, `smoker`, `region`) are stored as `object` — appropriate before encoding.  
> No dtype conversions were necessary.

## 5. Categorical & Numeric Value Checks

In [6]:
print('sex unique values   :', df['sex'].unique())
print('smoker unique values:', df['smoker'].unique())
print('region unique values:', df['region'].unique())

sex unique values   : ['female' 'male']
smoker unique values: ['yes' 'no']
region unique values: ['southwest' 'southeast' 'northwest' 'northeast']


In [7]:
print('age    min:', df['age'].min())
print('bmi    min:', df['bmi'].min())
print('children min:', df['children'].min())
print('charges min:', df['charges'].min())

age    min: 18
bmi    min: 15.96
children min: 0
charges min: 1121.8739


> **Result:** All categorical values are valid and consistent.  
> All numeric minimums are non-negative and within realistic ranges.  
> No invalid or inconsistent values found.

## 6. Before Preprocessing — Categorical Columns (Original State)

In [8]:
print('Categorical columns before encoding:')
df[['sex', 'smoker', 'region']].head()

Categorical columns before encoding:


,sex,smoker,region
0,female,yes,southwest
1,male,no,southeast
2,male,no,southeast
3,male,no,northwest
4,male,no,northwest


## 7. Categorical Encoding

Encode `sex`, `smoker`, and `region` using one-hot encoding (`pd.get_dummies`).  
`drop_first=True` removes one redundant dummy column per variable to avoid multicollinearity — suitable for linear regression.  

Columns created:
- `sex` → `sex_male` (1 = male, 0 = female)
- `smoker` → `smoker_yes` (1 = smoker, 0 = non-smoker)
- `region` → `region_northwest`, `region_southeast`, `region_southwest` (northeast is the dropped baseline)


In [9]:
df = pd.get_dummies(df, columns=['sex', 'smoker', 'region'], drop_first=True)
print('Encoding complete.')
print('New shape:', df.shape)

Encoding complete.
New shape: (1337, 9)


## 8. After Encoding — DataFrame Preview

In [10]:
df.head()

,age,bmi,children,charges,sex_male,smoker_yes,region_northwest,region_southeast,region_southwest
0,19,27.900,0,16884.92400,False,True,False,False,True
1,18,33.770,1,1725.55230,True,False,False,True,False
2,28,33.000,3,4449.46200,True,False,False,True,False
3,33,22.705,0,21984.47061,True,False,True,False,False
4,32,28.880,0,3866.85520,True,False,True,False,False


In [11]:
print('Data types after encoding:')
print(df.dtypes)

Data types after encoding:
age                   int64
bmi                 float64
children              int64
charges             float64
sex_male               bool
smoker_yes             bool
region_northwest       bool
region_southeast       bool
region_southwest       bool
dtype: object


In [12]:
print('Final column names:')
print(df.columns.tolist())

Final column names:
['age', 'bmi', 'children', 'charges', 'sex_male', 'smoker_yes', 'region_northwest', 'region_southeast', 'region_southwest']


## 9. Final Dataset

The target variable is `charges`.  
All remaining columns are input features.  
The dataset is now fully prepared for the modelling phase.

In [13]:
X = df.drop(columns=['charges'])
y = df['charges']

print('Features (X) shape:', X.shape)
print('Target  (y) shape:', y.shape)

Features (X) shape: (1337, 8)
Target  (y) shape: (1337,)


## 10. Final Checks

In [14]:
print('Final shape          :', df.shape)
print('Missing values total :', df.isnull().sum().sum())
print('Duplicate rows       :', df.duplicated().sum())
print('Final columns        :', df.columns.tolist())
print()
print('Missing values per column:')
print(df.isnull().sum())

Final shape          : (1337, 9)
Missing values total : 0
Duplicate rows       : 0
Final columns        : ['age', 'bmi', 'children', 'charges', 'sex_male', 'smoker_yes', 'region_northwest', 'region_southeast', 'region_southwest']

Missing values per column:
age                 0
bmi                 0
children            0
charges             0
sex_male            0
smoker_yes          0
region_northwest    0
region_southeast    0
region_southwest    0
dtype: int64


## 11. Preprocessing Summary

### Cleaning Performed

| Step | Action | Result |
|---|---|---|
| Missing values | Checked with `isnull().sum()` | None found — no treatment required |
| Duplicate rows | Checked with `duplicated().sum()` | **1,435 duplicates removed** (2,772 → 1,337 rows) |
| Data types | Checked all column dtypes | All valid — no conversion necessary |
| Categorical values | Inspected unique values | All valid — no inconsistencies found |
| Numeric values | Checked minimum values | All non-negative and realistic |
| Categorical encoding | `pd.get_dummies(drop_first=True)` | `sex`, `smoker`, `region` encoded |

### Final Dataset

- **Rows:** 1,337  
- **Columns:** 9  
- **Missing values:** 0  
- **Duplicate rows:** 0  
- **Encoded columns:** `sex_male`, `smoker_yes`, `region_northwest`, `region_southeast`, `region_southwest`

The dataset is ready for EDA and model training in the next phase.